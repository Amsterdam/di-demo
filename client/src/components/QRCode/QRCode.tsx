import React, { useState } from 'react';
import styled from 'styled-components';
import ReactMarkDown from 'react-markdown';

import content from '@services/content';

import { Button, Modal, themeSpacing, themeColor } from '@datapunt/asc-ui';
import { Linkedin, Close } from '@datapunt/asc-assets';
// @todo fix irma logo

export interface IProps {
  getSession: Function
}

const StyledButton = styled(Button)`
  margin-top: ${themeSpacing(6)};
  margin-bottom: ${themeSpacing(6)};
`;

const StyledH3 = styled.h3``;

const ModalWrapper = styled.div`
  padding: ${themeSpacing(0 5 5 5)};
  ul, ol {
    margin-top: 0;
  }
`;

const StyledHeader = styled.div`
  border-bottom: 1px solid ${themeColor('tint', 'level5')};
  display: flex;
  padding-left: ${themeSpacing(5)};
  justify-content: space-between;
  align-items: center;
`;

const CloseButton = styled(Button)`
  margin-right: ${themeSpacing(3)};
` ;

const CanvasWrapper = styled.div`
  display: flex;
  justify-content: center;
  padding: ${themeSpacing(3 0 5)};
` ;

const StyledOL = styled.ol`
  margin-top: 0;
`;


const QRCode: React.FC<IProps> = ({ getSession }) => {
  const [hasOverlay, setHasOverlay] = useState(false)

  const getQRSession = () => {
    setHasOverlay(true);
    typeof getSession === 'function' && getSession();
  }

  const closeModal = () => {
    setHasOverlay(false);
  }

  return (
    <div>
      <StyledButton onClick={getQRSession} variant="secondary" iconLeft={<Linkedin />}>
        <ReactMarkDown source={content.qrcode.knop} />
      </StyledButton>

      <Modal backdropOpacity={0.5} open={hasOverlay} onClose={closeModal}>
        <>
          <StyledHeader>
            <ReactMarkDown source={content.qrcode.title} renderers={{ heading: StyledH3 }} />
            <CloseButton size={30} variant="blank" icon={<Close />} onClick={closeModal} />
          </StyledHeader>
          <ModalWrapper>
            <ReactMarkDown source={content.qrcode.stappen} renderers={{ list: StyledOL }} />

            <CanvasWrapper>
              <canvas id="irma-qr" />
            </CanvasWrapper>
          </ModalWrapper>
        </>
      </Modal>

    </div>
  );
};

export default QRCode;
