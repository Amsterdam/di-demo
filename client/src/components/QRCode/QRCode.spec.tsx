import React from 'react';
import { screen, act, fireEvent } from '@testing-library/react';
import { setupMocks, wrappedRender } from '@test/utils';
import QRCode from './QRCode';

// Setup all the generic mocks
setupMocks();

// MockcreateIrmaSession
jest.mock('@services/createIrmaSession');

const getSession = jest.fn(async () => {
    return { data: 'Test response' };
});

describe('QR Code', () => {
    it('should render the initial QR Code button, and not the modal', async () => {
        // Render QR Code
        await act(async () => await wrappedRender(<QRCode label="Test" getSession={getSession} />));

        const QRCodeButton: HTMLElement = await screen.findByTestId('qrCodeButton');
        expect(QRCodeButton).toBeInTheDocument();

        const QRCodeModal: HTMLElement = await screen.queryByTestId('qrCodeModal');
        expect(QRCodeModal).toBeNull();
    });

    it('Clicking the QR Code button should show the modal and call getSession', async () => {
        // Render QR Code
        await act(async () => await wrappedRender(<QRCode label="Test" getSession={getSession} />));

        const QRCodeButton: HTMLElement = await screen.findByTestId('qrCodeButton');

        fireEvent.click(QRCodeButton);

        const QRCodeModal: HTMLElement = await screen.findByTestId('qrCodeModal');
        expect(QRCodeModal).toMatchSnapshot();

        expect(getSession.mock.calls.length).toBe(1);
    });
});
