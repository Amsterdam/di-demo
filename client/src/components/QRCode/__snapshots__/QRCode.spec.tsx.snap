// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QR Code Clicking the QR Code button should show the modal and call getSession 1`] = `
.c7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  width: 20px;
  height: 20px;
  box-sizing: content-box;
  -webkit-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
}

.c7 > svg {
  width: 20px;
  height: 20px;
}

.c17:not(:disabled).c3:focus .sc-gsTCUz,
.c17:not(:disabled).c17:hover .sc-gsTCUz {
  border-left-color: #b90000;
}

.c17 .c6 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4 {
  height: 44px;
  white-space: nowrap;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  border: none;
  cursor: pointer;
  font-size: 16px;
  font-weight: 700;
  line-height: 20px;
  padding: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 30px;
  width: 30px;
  -webkit-transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
  transition: color 0.1s ease-in-out,background-color 0.1s ease-in-out;
  background-color: #ffffff;
}

.c4 svg circle,
.c4 svg rect,
.c4 svg polygon,
.c4 svg path {
  fill: #000000;
}

.c4:hover {
  background-color: #E6E6E6;
}

.c4::after {
  content: '';
  display: block;
  min-height: inherit;
  font-size: 0;
}

.c4 .c6 {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c4:disabled {
  cursor: default;
  outline: none;
  border: none;
  color: #B4B4B4;
  background-color: #E6E6E6;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c4:disabled svg circle,
.c4:disabled svg rect,
.c4:disabled svg polygon,
.c4:disabled svg path {
  fill: #B4B4B4;
}

.c1 {
  background-color: #ffffff;
  overflow: auto;
  width: 100%;
  max-width: 620px;
  max-height: 75%;
  position: fixed;
  top: 50%;
  left: 50%;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
  pointer-events: all;
  z-index: 21;
}

.c1 .sc-GTWni {
  min-height: 54px;
}

.c0 {
  display: grid;
  height: 100vh;
  margin: 0;
  place-items: center center;
  pointer-events: none;
}

.c9 {
  line-height: 24px;
  margin-top: 0;
}

.c9 li p {
  margin-left: 16px;
}

.c11 {
  position: absolute;
  width: 65px;
  height: 65px;
  top: 119px;
  left: 119px;
}

.c8 {
  padding: 0 20px 20px 20px;
}

.c2 {
  border-bottom: 1px solid #767676;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding-left: 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c5 {
  margin-right: 12px;
}

.c10 {
  position: relative;
  width: 300px !important;
  height: 300px !important;
  margin: 50px auto;
}

.c16 {
  width: 300px !important;
  height: 300px !important;
  background-color: transparent;
}

.c16 .irma-web-header {
  display: none;
}

.c16 .irma-web-content {
  margin: 0;
}

.c16 .irma-web-qr-canvas {
  width: 300px !important;
  height: 300px !important;
}

.c12 {
  width: 40px;
  height: 40px;
  position: absolute;
  top: -20px;
  left: -20px;
}

.c13 {
  width: 40px;
  height: 40px;
  position: absolute;
  top: -20px;
  right: -20px;
}

.c14 {
  width: 40px;
  height: 40px;
  position: absolute;
  bottom: -20px;
  right: -20px;
}

.c15 {
  width: 40px;
  height: 40px;
  position: absolute;
  bottom: -20px;
  left: -20px;
}

<div
  class="c0"
  data-testid="qrCodeModal"
>
  <div
    class="c1"
    role="dialog"
  >
    <div
      class="c2"
    >
      <h3
        class=""
      >
        Maak uzelf bekend
      </h3>
      <button
        class="c3 c4 c5"
        size="30"
      >
        <span
          class="c6 c7"
          size="20"
        >
          <svg
            focusable="false"
            viewBox="0 0 32 32"
          >
            <path
              d="M29.333 5.47L26.53 2.668 16 13.187 5.47 2.666 2.668 5.47 13.187 16 2.666 26.53l2.804 2.803L16 18.813l10.53 10.52 2.803-2.804L18.813 16z"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
    </div>
    <div
      class="c8"
    >
      <p>
        Doorloop de volgende stappen: 
      </p>
      <ol
        class="c9"
        start="1"
      >
        <li>
          <p>
            Scan de QR-code hieronder met uw IRMA-app.  
          </p>
        </li>
        <li>
          <p>
            Kies in uw IRMA-app of u de gevraagde gegevens wilt doorgeven om in te loggen op Mijn Amsterdam.
          </p>
        </li>
      </ol>
      <div
        class="c10"
      >
        <img
          class="c11"
          src="/assets/irma_logo.svg"
        />
        <img
          class="c12"
          src="/assets/qr-top-left.svg"
        />
        <img
          class="c13"
          src="/assets/qr-top-right.svg"
        />
        <img
          class="c14"
          src="/assets/qr-bottom-right.svg"
        />
        <img
          class="c15"
          src="/assets/qr-bottom-left.svg"
        />
        <div
          class="c16"
          id="irma-qr"
        />
      </div>
    </div>
  </div>
</div>
`;
